## Anime Recommender

This is a simple anime recommender using cosine similarity based on the anime's genres. The data used was scraped from MyAnimeList using a custom [web scraper](https://github.com/rmagtanong/myanimelist-scraper) I created using Scrapy.

In [1]:
import pandas as pd
import numpy as np
from sklearn.metrics.pairwise import cosine_similarity

In [2]:
df = pd.read_csv('anime.csv')

In [4]:
df.head(10)

anime_id                                               name  rating  \
0        33                               Kenpuu Denki Berserk    8.54   
1     21939                                 Mushishi Zoku Shou    8.70   
2     38889                            Kono Oto Tomare! Part 2    8.43   
3        30                            Neon Genesis Evangelion    8.34   
4      2685                         Tsubasa: Tokyo Revelations    8.29   
5      2966                              Ookami to Koushinryou    8.23   
6     33051  Mobile Suit Gundam: Iron-Blooded Orphans 2nd S...    8.23   
7     32843                        Senki Zesshou Symphogear XV    8.19   
8     39198                                    Kanata no Astra    8.10   
9      2164                                        Dennou Coil    8.06   

                                              genres  users_watched  \
0  Action,Adventure,Drama,Fantasy,Horror,Supernat...         548221   
1  Adventure,Fantasy,Mystery,Slice of Life,Supern...         272380   
2  Drama,Romance,Music,Performing Arts,School,Sho...         141645   
3  Action,Avant Garde,Drama,Sci-Fi,Mecha,Psycholo...        1577154   
4     Action,Adventure,Drama,Fantasy,Romance,Shounen          90097   
5    Adventure,Fantasy,Romance,Adult Cast,Historical         748249   
6                    Action,Drama,Sci-Fi,Mecha,Space         123153   
7                 Action,Sci-Fi,Idols (Female),Music          34642   
8    Adventure,Mystery,Sci-Fi,Space,Survival,Shounen         264777   
9              Adventure,Comedy,Drama,Mystery,Sci-Fi         136460   

   users_rated                                                url  
0       293480  https://myanimelist.net/anime/33/Kenpuu_Denki_...  
1       107898  https://myanimelist.net/anime/21939/Mushishi_Z...  
2        75844  https://myanimelist.net/anime/38889/Kono_Oto_T...  
3       940938  https://myanimelist.net/anime/30/Neon_Genesis_...  
4        43459  https://myanimelist.net/anime/2685/Tsubasa__To...  
5       364345  https://myanimelist.net/anime/2966/Ookami_to_K...  
6        69445  https://myanimelist.net/anime/33051/Mobile_Sui...  
7        11982  https://myanimelist.net/anime/32843/Senki_Zess...  
8       129276  https://myanimelist.net/anime/39198/Kanata_no_...  
9        40977     https://myanimelist.net/anime/2164/Dennou_Coil

In [5]:
df.shape

(10049, 7)

In [6]:
df.isnull().sum()

anime_id          0
name              0
rating            0
genres           18
users_watched     0
users_rated       0
url               0
dtype: int64

In [7]:
df[df['genres'].isnull()]

anime_id                                               name  rating  \
1825     50549                                             Bubble    7.31   
2084      6399                      Higashi no Eden: Falling Down    7.37   
2508     51308                                               Yoku    7.49   
2577     51230                         Taikutsu wo Saien Shinaide    7.14   
4435     32468                                            Nirvana    6.79   
4635     32700                                      Heart Realize    6.83   
5098     50288                                         Cinderella    6.93   
5830     50352                                 Peko Random Brain!    6.73   
6014     50964                                         Koufukuron    6.46   
7470      6368                                   Legend of Regios    6.42   
7505     51092                                         Yurei Deco    6.42   
7699      9911                                Houkago no Pleiades    6.14   
8741     27997  Kaitou Tenshi Twin Angel: Kyun Kyun☆Tokimeki P...    6.02   
8855      3805                           Kaitou Tenshi Twin Angel    6.03   
8927     12121                                     Towa no Kizuna    6.05   
8939     15407                           Magical Suite Prism Nana    6.05   
9047     10217  Kaitou Tenshi Twin Angel: Kyun Kyun☆Tokimeki P...    6.07   
9802     38341                                           Bai Niao    5.89   

     genres  users_watched  users_rated  \
1825    NaN         130589        64702   
2084    NaN          13639         7590   
2508    NaN            826          488   
2577    NaN            626          357   
4435    NaN           7869         3541   
4635    NaN           7648         3496   
5098    NaN           1354          798   
5830    NaN            473          293   
6014    NaN            542          289   
7470    NaN           9982         2682   
7505    NaN          12349         1623   
7699    NaN          17236         8666   
8741    NaN           1392          299   
8855    NaN           5722         1942   
8927    NaN           4316         2242   
8939    NaN           9102          178   
9047    NaN           8066         2432   
9802    NaN            539          184   

                                                    url  
1825         https://myanimelist.net/anime/50549/Bubble  
2084  https://myanimelist.net/anime/6399/Higashi_no_...  
2508           https://myanimelist.net/anime/51308/Yoku  
2577  https://myanimelist.net/anime/51230/Taikutsu_w...  
4435        https://myanimelist.net/anime/32468/Nirvana  
4635  https://myanimelist.net/anime/32700/Heart_Realize  
5098     https://myanimelist.net/anime/50288/Cinderella  
5830  https://myanimelist.net/anime/50352/Peko_Rando...  
6014     https://myanimelist.net/anime/50964/Koufukuron  
7470  https://myanimelist.net/anime/6368/Legend_of_R...  
7505     https://myanimelist.net/anime/51092/Yurei_Deco  
7699  https://myanimelist.net/anime/9911/Houkago_no_...  
8741  https://myanimelist.net/anime/27997/Kaitou_Ten...  
8855  https://myanimelist.net/anime/3805/Kaitou_Tens...  
8927  https://myanimelist.net/anime/12121/Towa_no_Ki...  
8939  https://myanimelist.net/anime/15407/Magical_Su...  
9047  https://myanimelist.net/anime/10217/Kaitou_Ten...  
9802       https://myanimelist.net/anime/38341/Bai_Niao

In [9]:
100*df.isnull().sum()/len(df)

anime_id         0.000000
name             0.000000
rating           0.000000
genres           0.179122
users_watched    0.000000
users_rated      0.000000
url              0.000000
dtype: float64

#### We will use the genres to generate our recommender, so we can't have missing values. We can drop the entries since missing data only accounts for 0.179122% of our data

In [10]:
df = df.dropna()
df.isnull().sum()

anime_id         0
name             0
rating           0
genres           0
users_watched    0
users_rated      0
url              0
dtype: int64

#### Let's drop anime_id and users_rated since we don't need them

In [11]:
df.drop(['anime_id', 'users_rated'], axis=1, inplace=True)

In [12]:
df.head()

name  rating  \
0        Kenpuu Denki Berserk    8.54   
1          Mushishi Zoku Shou    8.70   
2     Kono Oto Tomare! Part 2    8.43   
3     Neon Genesis Evangelion    8.34   
4  Tsubasa: Tokyo Revelations    8.29   

                                              genres  users_watched  \
0  Action,Adventure,Drama,Fantasy,Horror,Supernat...         548221   
1  Adventure,Fantasy,Mystery,Slice of Life,Supern...         272380   
2  Drama,Romance,Music,Performing Arts,School,Sho...         141645   
3  Action,Avant Garde,Drama,Sci-Fi,Mecha,Psycholo...        1577154   
4     Action,Adventure,Drama,Fantasy,Romance,Shounen          90097   

                                                 url  
0  https://myanimelist.net/anime/33/Kenpuu_Denki_...  
1  https://myanimelist.net/anime/21939/Mushishi_Z...  
2  https://myanimelist.net/anime/38889/Kono_Oto_T...  
3  https://myanimelist.net/anime/30/Neon_Genesis_...  
4  https://myanimelist.net/anime/2685/Tsubasa__To...

#### We will recommend based solely on genres, so let's explore the column

In [13]:
df['genres'][0]

'Action,Adventure,Drama,Fantasy,Horror,Supernatural,Gore,Military,Mythology,Seinen'

In [14]:
type(df['genres'][0])

str

#### Create dummies for each Genre

In [15]:
df = pd.concat([df, df['genres'].str.get_dummies(sep=',')], axis=1)

In [16]:
df.head()

name  rating  \
0        Kenpuu Denki Berserk    8.54   
1          Mushishi Zoku Shou    8.70   
2     Kono Oto Tomare! Part 2    8.43   
3     Neon Genesis Evangelion    8.34   
4  Tsubasa: Tokyo Revelations    8.29   

                                              genres  users_watched  \
0  Action,Adventure,Drama,Fantasy,Horror,Supernat...         548221   
1  Adventure,Fantasy,Mystery,Slice of Life,Supern...         272380   
2  Drama,Romance,Music,Performing Arts,School,Sho...         141645   
3  Action,Avant Garde,Drama,Sci-Fi,Mecha,Psycholo...        1577154   
4     Action,Adventure,Drama,Fantasy,Romance,Shounen          90097   

                                                 url  Action  Adult Cast  \
0  https://myanimelist.net/anime/33/Kenpuu_Denki_...       1           0   
1  https://myanimelist.net/anime/21939/Mushishi_Z...       0           1   
2  https://myanimelist.net/anime/38889/Kono_Oto_T...       0           0   
3  https://myanimelist.net/anime/30/Neon_Genesis_...       1           0   
4  https://myanimelist.net/anime/2685/Tsubasa__To...       1           0   

   Adventure  Anthropomorphic  Avant Garde  ...  Super Power  Supernatural  \
0          1                0            0  ...            0             1   
1          1                0            0  ...            0             1   
2          0                0            0  ...            0             0   
3          0                0            1  ...            0             0   
4          1                0            0  ...            0             0   

   Survival  Suspense  Team Sports  Time Travel  Vampire  Video Game  \
0         0         0            0            0        0           0   
1         0         0            0            0        0           0   
2         0         0            0            0        0           0   
3         0         0            0            0        0           0   
4         0         0            0            0        0           0   

   Visual Arts  Workplace  
0            0          0  
1            0          0  
2            0          0  
3            0          0  
4            0          0  

[5 rows x 79 columns]

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10031 entries, 0 to 10048
Data columns (total 79 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   name               10031 non-null  object 
 1   rating             10031 non-null  float64
 2   genres             10031 non-null  object 
 3   users_watched      10031 non-null  int64  
 4   url                10031 non-null  object 
 5   Action             10031 non-null  int64  
 6   Adult Cast         10031 non-null  int64  
 7   Adventure          10031 non-null  int64  
 8   Anthropomorphic    10031 non-null  int64  
 9   Avant Garde        10031 non-null  int64  
 10  Award Winning      10031 non-null  int64  
 11  Boys Love          10031 non-null  int64  
 12  CGDCT              10031 non-null  int64  
 13  Childcare          10031 non-null  int64  
 14  Combat Sports      10031 non-null  int64  
 15  Comedy             10031 non-null  int64  
 16  Crossdressing      100

In [18]:
anime_features = df.loc[:, 'Action':].copy()

In [19]:
anime_features.head()

Action  Adult Cast  Adventure  Anthropomorphic  Avant Garde  Award Winning  \
0       1           0          1                0            0              0   
1       0           1          1                0            0              0   
2       0           0          0                0            0              0   
3       1           0          0                0            1              0   
4       1           0          1                0            0              0   

   Boys Love  CGDCT  Childcare  Combat Sports  ...  Super Power  Supernatural  \
0          0      0          0              0  ...            0             1   
1          0      0          0              0  ...            0             1   
2          0      0          0              0  ...            0             0   
3          0      0          0              0  ...            0             0   
4          0      0          0              0  ...            0             0   

   Survival  Suspense  Team Sports  Time Travel  Vampire  Video Game  \
0         0         0            0            0        0           0   
1         0         0            0            0        0           0   
2         0         0            0            0        0           0   
3         0         0            0            0        0           0   
4         0         0            0            0        0           0   

   Visual Arts  Workplace  
0            0          0  
1            0          0  
2            0          0  
3            0          0  
4            0          0  

[5 rows x 74 columns]

In [20]:
cos_sim = cosine_similarity(anime_features.values, anime_features.values)

In [21]:
cos_sim

array([[1.        , 0.42163702, 0.12909944, ..., 0.36514837, 0.        ,
        0.2236068 ],
       [0.42163702, 1.        , 0.        , ..., 0.19245009, 0.        ,
        0.23570226],
       [0.12909944, 0.        , 1.        , ..., 0.23570226, 0.        ,
        0.        ],
       ...,
       [0.36514837, 0.19245009, 0.23570226, ..., 1.        , 0.        ,
        0.40824829],
       [0.        , 0.        , 0.        , ..., 0.        , 1.        ,
        0.35355339],
       [0.2236068 , 0.23570226, 0.        , ..., 0.40824829, 0.35355339,
        1.        ]])

In [22]:
cos_sim.shape

(10031, 10031)

In [23]:
anime_index = pd.Series(df.index, index=df.name).drop_duplicates()

In [24]:
anime_index

name
Kenpuu Denki Berserk                           0
Mushishi Zoku Shou                             1
Kono Oto Tomare! Part 2                        2
Neon Genesis Evangelion                        3
Tsubasa: Tokyo Revelations                     4
                                           ...  
Ichirin-sha                                10044
Hitojichi Koukan feat. Hatsune Miku        10045
Highschool Aurabuster: Hikari no Mezame    10046
Heybot!                                    10047
Hello Kitty no Yuki no Joou                10048
Length: 10031, dtype: int64

In [41]:
def recommend_anime(anime_name, similarity=cos_sim):
    
    # get index of anime input
    index = anime_index[anime_name]
    
    # get cosine similarity scores from input anime
    similarity_scores = list(enumerate(cos_sim[index]))
    
    # sort list from highest score to lowest
    similarity_scores = sorted(similarity_scores, key=lambda x: x[1], reverse=True)
    
    # get only 10 highest similarity scores
    similarity_scores = similarity_scores[0:11]

    # get indices of top 10 animes
    anime_indices = [anime_index[0] for anime_index in similarity_scores]

    # get anime info from index
    recommendations = df[['name', 'genres', 'rating', 'users_watched', 'url']].iloc[anime_indices].drop(index)
    
    return recommendations

In [42]:
recommend_anime('Cowboy Bebop')

name  \
677                     Cowboy Bebop: Tengoku no Tobira   
139   Ginga Eiyuu Densetsu: Waga Yuku wa Hoshi no Ta...   
1400                                        Space Cobra   
3402                                Cobra The Animation   
4026                              Space Adventure Cobra   
4826                    Cobra The Animation: Time Drive   
7515                                  Space Cobra Pilot   
2277                    Cowboy Bebop: Yose Atsume Blues   
5074                    Seihou Bukyou Outlaw Star Pilot   
5181                               Wonder Beat Scramble   

                                        genres  rating  users_watched  \
677             Action,Sci-Fi,Adult Cast,Space    8.38         337711   
139    Action,Sci-Fi,Adult Cast,Military,Space    7.90          35385   
1400  Action,Adventure,Sci-Fi,Adult Cast,Space    7.68          19710   
3402  Action,Adventure,Sci-Fi,Adult Cast,Space    7.09           8254   
4026  Action,Adventure,Sci-Fi,Adult Cast,Space    7.09           9292   
4826  Action,Adventure,Sci-Fi,Adult Cast,Space    6.87           4033   
7515  Action,Adventure,Sci-Fi,Adult Cast,Space    6.42           1997   
2277                   Sci-Fi,Adult Cast,Space    7.42          44565   
5074                       Action,Sci-Fi,Space    6.93           6381   
5181                       Action,Sci-Fi,Space    6.61           1012   

                                                    url  
677   https://myanimelist.net/anime/5/Cowboy_Bebop__...  
139   https://myanimelist.net/anime/3014/Ginga_Eiyuu...  
1400     https://myanimelist.net/anime/2451/Space_Cobra  
3402  https://myanimelist.net/anime/5032/Cobra_The_A...  
4026  https://myanimelist.net/anime/2452/Space_Adven...  
4826  https://myanimelist.net/anime/5031/Cobra_The_A...  
7515  https://myanimelist.net/anime/5742/Space_Cobra...  
2277  https://myanimelist.net/anime/4037/Cowboy_Bebo...  
5074  https://myanimelist.net/anime/4650/Seihou_Buky...  
5181  https://myanimelist.net/anime/4083/Wonder_Beat...

In [44]:
recommend_anime('Mob Psycho 100')

name  \
822                                   Mob Psycho 100 II   
1267  Mob Psycho 100: Dai Ikkai Rei toka Soudansho I...   
1913  Mob Psycho 100: Reigen - Shirarezaru Kiseki no...   
173                       Bungou Stray Dogs: Dead Apple   
1701                      Yozakura Quartet: Hana no Uta   
2134                     Yozakura Quartet: Hoshi no Umi   
2498                    Yozakura Quartet: Tsuki ni Naku   
4494                                   Yozakura Quartet   
75                                  Durarara!! Specials   
1116                                 K: Return of Kings   

                                              genres  rating  users_watched  \
822           Action,Comedy,Supernatural,Super Power    8.81        1260624   
1267          Action,Comedy,Supernatural,Super Power    7.63          80548   
1913          Action,Comedy,Supernatural,Super Power    7.33          81922   
173   Action,Comedy,Mystery,Supernatural,Super Power    7.91         222469   
1701  Action,Comedy,Supernatural,Super Power,Shounen    7.47         117386   
2134  Action,Comedy,Supernatural,Super Power,Shounen    7.39          37057   
2498  Action,Comedy,Supernatural,Super Power,Shounen    7.48          27394   
4494  Action,Comedy,Supernatural,Super Power,Shounen    6.80         143725   
75                        Action,Comedy,Supernatural    7.86         161364   
1116                 Action,Supernatural,Super Power    7.57         307078   

                                                    url  
822   https://myanimelist.net/anime/37510/Mob_Psycho...  
1267  https://myanimelist.net/anime/39651/Mob_Psycho...  
1913  https://myanimelist.net/anime/36616/Mob_Psycho...  
173   https://myanimelist.net/anime/34944/Bungou_Str...  
1701  https://myanimelist.net/anime/18497/Yozakura_Q...  
2134  https://myanimelist.net/anime/8457/Yozakura_Qu...  
2498  https://myanimelist.net/anime/18499/Yozakura_Q...  
4494  https://myanimelist.net/anime/4548/Yozakura_Qu...  
75    https://myanimelist.net/anime/8408/Durarara_Sp...  
1116  https://myanimelist.net/anime/27991/K__Return_...

In [45]:
recommend_anime('Great Teacher Onizuka')

name  \
9495                                  Rokudenashi Blues   
161                         Sakigake!! Cromartie Koukou   
2095                  Araburu Kisetsu no Otome-domo yo.   
2249                                            Gokusen   
7039                             Rokudenashi Blues 1993   
192   SKET Dance: Imouto no Nayami ni Nayamu Ani ni ...   
355                                  Kyou kara Ore wa!!   
503                                          SKET Dance   
3691    Yankee-kun na Yamada-kun to Megane-chan to Majo   
5024                               Houkago no Ouji-sama   

                                                 genres  rating  \
9495  Comedy,Drama,Sports,Combat Sports,Delinquents,...    5.81   
161         Comedy,Delinquents,Gag Humor,School,Shounen    7.91   
2095                Comedy,Drama,Romance,School,Shounen    7.37   
2249  Comedy,Drama,Slice of Life,Delinquents,Organiz...    7.41   
7039  Action,Comedy,Drama,Sports,Combat Sports,Delin...    6.32   
192                               Comedy,School,Shounen    7.93   
355                          Comedy,Delinquents,Shounen    8.06   
503                               Comedy,School,Shounen    8.22   
3691                              Comedy,School,Shounen    7.01   
5024                              Comedy,School,Shounen    6.92   

      users_watched                                                url  
9495           2038  https://myanimelist.net/anime/9077/Rokudenashi...  
161          128069  https://myanimelist.net/anime/114/Sakigake_Cro...  
2095         291147  https://myanimelist.net/anime/38753/Araburu_Ki...  
2249          41272          https://myanimelist.net/anime/242/Gokusen  
7039           1112  https://myanimelist.net/anime/9078/Rokudenashi...  
192           20542  https://myanimelist.net/anime/16395/SKET_Dance...  
355           29704  https://myanimelist.net/anime/851/Kyou_kara_Or...  
503          211074      https://myanimelist.net/anime/9863/SKET_Dance  
3691          23690  https://myanimelist.net/anime/30641/Yankee-kun...  
5024           2859  https://myanimelist.net/anime/24457/Houkago_no...

Take this example, from [Kaguya-sama Season 3's](https://myanimelist.net/anime/43608/Kaguya-sama_wa_Kokurasetai__Ultra_Romantic) page:

<img align="left" src="kaguya.png">


Compare to [Cowboy Bebop](https://myanimelist.net/anime/1/Cowboy_Bebop) (no Demographic):

<img align="left" src="bebop.png">


Compare to [Made in Abyss](https://myanimelist.net/anime/34599/Made_in_Abyss) (no Themes/Demographic):

<img align="left" src="madeinabyss.png">

Ideally genres, themes, and demographic would have their own fields. I could not separate them due to limitations with MyAnimeList's website design. Also, as seen above, not all anime have complete fields for Genres, Themes, and Demographic. 

The recommender could be improved with better data, which could be achieved by:

1. Separating Genres, Themes, and Demographic into their own columns using more advanced scraping methods
2. Using feature engineering to separate the columns
3. Scraping from a website with more complete data